ci/ci

```diff
@@ -0,0 +1,344 @@
+        return
+    bad_images = (bad_images.split("'")[1]).replace("\\n", " ")
+    log_info("The following untagged images will be removed -> " + bad_images)
+    cmd = 'docker rmi ' + bad_images
+    run_cmd(cmd)
+
+
+def find_ci():
+    cur_dir = os.path.dirname(os.path.abspath(__file__))
+    return os.path.dirname(cur_dir)
+
+
+def tail_file(build_log_path):
+    with open(build_log_path, 'rb') as f:
+        f.seek(-2, os.SEEK_END)
+        while f.read(1) != b'\n':
+            f.seek(-2, os.SEEK_CUR)
+        return f.readline().decode()
+
+
+
+def check_for_existing_image():
+    env_dir = os.path.join(pbs_dirname, '.env')
+    val = read_from_file(env_dir)
+    if val != '':
+        val = val.split('=')[1]
+        if val.find('_ci_pbs') != -1:
+            val = val[:-7].replace('_', ':')
+    else:
+        val = 'centos:7'
+    search_str = val.replace(":", "_")
+    search_str += '_ci_pbs'
+    cmd = 'docker images -q ' + search_str
+    search_result = run_cmd(cmd,True)
+    if search_result != "b''":
+        write_to_file(env_dir, "CONTAINER_PLATFORM="+search_str)
+        return True
+    else:
+        write_to_file(env_dir, "CONTAINER_PLATFORM="+val)
+        return False
+
+
+def check_prerequisites():
+    # This function will check wether docker docker-compose commands are available
+    # or a  container is running that has pbs
+    cmd = "where" if platform.system() == "Windows" else "which"
+
+    try:
+        subprocess.run([cmd, "docker"], stdout=subprocess.DEVNULL)
+    except:
+        log_error("docker not found in PATH")
+
+    try:
+        subprocess.run([cmd, "docker-compose"], stdout=subprocess.DEVNULL)
+    except:
+        log_error("docker-compose not found in PATH")
+
+    try:
+        docker_container_process = subprocess.run(
+            ["docker-compose", "ps", "pbs-build"], stdout=subprocess.PIPE)
+        docker_container = str(docker_container_process.stdout)
+        if docker_container.find('Up') == -1:
+            log_error("No running service found")
+            try:
+                log_info('Attempting to start container')
+                os.chdir(pbs_dirname)
+                subprocess.run(["docker-compose", "down"],
+                               stdout=subprocess.DEVNULL)
+                image_check = check_for_existing_image()
+                subprocess.run(["docker-compose", "up", "-d"])
+                if image_check == True:
+                    return 0
+                log_info('Waiting for container build to complete ')
+                build_log_path = os.path.join(pbs_dirname, 'logs', 'build')
+                log_info("You can check the logs in " + build_log_path)
+                # wait for build to complete and commit newly built container
+                timer = 1
+                while True:
+                    time.sleep(30)
+                    last_line = tail_file(build_log_path)
+                    log_info("Get current status ...... " + last_line)
```

Comment:
i am not sure, if this log is reporting progress to the user.

Reply:
This is not reporting progress exactly to the user, for that the user must go to the log file, however it gives some idea as to what is the current status of the build as opposed to a blinking cursor on the screen for some 10 mins.
```
14:50:12 ---> Waiting for container build to complete 
14:50:12 ---> Build logs can be found in /home/sanidhya/instantCI/pbspro/ci/logs/build
14:50:42 ---> Get current status ...... Public key for device-mapper-1.02.158-2.el7_7.2.x86_64.rpm is not installed

14:51:12 ---> Get current status ......   Cleanup    : bash-4.2.46-31.el7.x86_64                                124/130
14:51:42 ---> Get current status ......  * updates: centos.myfahim.com

14:52:12 ---> Get current status ......   Installing : openssh-clients-7.4p1-21.el7.x86_64                        38/57
14:52:42 ---> Get current status ...... Downloading packages:

14:53:12 ---> Get current status ......   Verifying  : libXau-devel-1.0.8-2.1.el7.x86_64                          73/73
14:53:42 ---> Get current status ...... ### First time build is complete ###

14:53:42 ---> build is now complete
14:53:43 ---> docker commit b6136476d886 centos_7_ci_pbs:latest
14:53:50 --->  done start container 
```
Any alternatives would be welcome as well, do you have better suggestions?

Reply:
hmmm, what you say is like reporting progress...
let us remove "Get" from the log...